helix-core/src/main/java/org/apache/helix/task/TaskRunner.java

```diff
@@ -180,15 +182,16 @@ private void signalStarted() {
    * Requests the controller for a state transition.
    * @param state The state transition that is being requested.
    */
-  private void requestStateTransition(TaskPartitionState state) {
-    boolean success =
-        setRequestedState(_manager.getHelixDataAccessor(), _instance, _sessionId, _taskName,
-            _taskPartition, state);
-    if (!success) {
-      LOG.error(String
-          .format(
-              "Failed to set the requested state to %s for instance %s, session id %s, task partition %s.",
-              state, _instance, _sessionId, _taskPartition));
+  private void updateCurrentState(TaskPartitionState state) {
+    synchronized (_stateModel) {
+      if (setZKCurrentState(_manager.getHelixDataAccessor(), _instance, _sessionId, _taskName,
```

Comment:
In postHandleMessage, we set statemodel before update current state. Shall we do the same here? Or you have some other thought?